dask_cuda/explicit_comms/dataframe/shuffle.py

```diff
@@ -241,118 +279,80 @@ def shuffle(
     Developer Notes
     ---------------
     The implementation consist of three steps:
-      (a) Extend the dask graph of `df` with a call to `shuffle_group()` for each
-          dataframe partition and submit the graph.
+      (a) Stage the partitions of `df` on all workers and then cancel them
+          thus at this point the Dask Scheduler doesn't know about any of the
+          the partitions.
       (b) Submit a task on each worker that shuffle (all-to-all communicate)
-          the groups from (a) and return a list of dataframe-partitions.
+          the staged partitions and return a list of dataframe-partitions.
       (c) Submit a dask graph that extract (using `getitem()`) individual
           dataframe-partitions from (b).
     """
     c = comms.default_comms()
 
-    # As default we preserve number of partitions
+    # The ranks of the output workers
+    ranks = list(range(len(c.worker_addresses)))
+
+    # By default, we preserve number of partitions
     if npartitions is None:
         npartitions = df.npartitions
 
-    # Step (a): partition/group each dataframe-partition
+    # Step (a):
+    df = df.persist()  # Make sure optimizations are apply on the existing graph
+    wait(df)  # Make sure all keys has been materialized on workers
     name = (
-        "explicit-comms-shuffle-group-"
+        "explicit-comms-shuffle-"
         f"{tokenize(df, column_names, npartitions, ignore_index)}"
     )
-    df = df.persist()  # Making sure optimizations are apply on the existing graph
-    dsk = dict(df.__dask_graph__())
-    output_keys = []
-    for input_key in df.__dask_keys__():
-        output_key = (name, input_key[1])
-        dsk[output_key] = (
-            shuffle_group,
-            input_key,
+    df_meta: DataFrame = df._meta
+
+    # Stage all keys of `df` on the workers and cancel them, which makes it possible
+    # for the shuffle to free memory as the partitions of `df` are consumed.
+    rank_to_inkeys = c.stage_keys(name=name, keys=df.__dask_keys__())
+    c.client.cancel(df)  # Notice, since `df` has been staged, nothing is freed here.
```

Comment:
This I do not understand. We just copy some keys around here as far as I can tell (and don't persist anything), so I'm not sure what this even does.

Reply:
Added more doc in 7ea9f6fb32c144ae922ad515bad8152c69888fd2